### PR TITLE
Gather optimization info through the DAC

### LINKED ISSUE #89474: Have AreOptimizationsDisabled to go through DAC
Rather than using an IPCEvent to see if code optimizations are disabled, go through the DAC to collect this information.

### PR DESCRIPTION
**Pull request by mikelle-rogers**
Fixes #89474 
Added DacDbiInterfaceImpl::AreOptimizationsDisabled

### CODE REVIEW

File: src/coreclr/debug/daccess/dacdbiimpl.cpp [hunk 1]
```diff
@@ -7407,6 +7407,24 @@ HRESULT DacDbiInterfaceImpl::GetReJitInfo(VMPTR_MethodDesc vmMethod, CORDB_ADDRE
...
+HRESULT DacDbiInterfaceImpl::AreOptimizationsDisabled(VMPTR_Module vmModule, mdMethodDef methodTk, OUT BOOL* pOptimizationsDisabled)
+{
+    DD_ENTER_MAY_THROW;
+    PTR_Module pModule = vmModule.GetDacPtr();
+    if (pModule == NULL || pOptimizationsDisabled == NULL || TypeFromToken(methodTk) != mdtMethodDef)
+    {
+        return E_INVALIDARG;
+    }
+    {
+        CodeVersionManager::LockHolder codeVersioningLockHolder;
```
Comment (hoyosjs) [inline src/coreclr/debug/daccess/dacdbiimpl.cpp:7419]: These should follow suit of other methods here and guard on FEATURE_REJIT (just for consistency, as some point we might want to remove). Also, the lock is not necessary in the DAC

Comment (mikelle-rogers) [reply src/coreclr/debug/daccess/dacdbiimpl.cpp:7420]: Let me know if my latest commit is what you meant.

File: src/coreclr/debug/di/rsfunction.cpp [hunk 2]
```diff
@@ -673,27 +673,12 @@ HRESULT CordbFunction::AreOptimizationsDisabled(BOOL *pOptimizationsDisabled)
...
-    lockHolder.Acquire();
-
-    _ASSERTE(event.type == DB_IPCE_IS_OPTS_DISABLED_RESULT);
-    
-    *pOptimizationsDisabled = event.IsOptsDisabledData.value;
-
-    return event.hr;;
+    EX_TRY
+    {
+        hr = GetProcess()->GetDAC()->AreOptimizationsDisabled(GetModule()->m_vmModule, m_MDToken, pOptimizationsDisabled);
```
[the remark this answers is not in the document]
Comment (mikelle-rogers) [reply src/coreclr/debug/di/rsfunction.cpp]: I changed this, and now it no longer builds...cannot convert argument 1 from 'CordbModule *' to 'VMPTR_Module'

File: src/coreclr/debug/daccess/dacdbiimpl.cpp [hunk 3]
```diff
@@ -7407,6 +7407,28 @@ HRESULT DacDbiInterfaceImpl::GetReJitInfo(VMPTR_MethodDesc vmMethod, CORDB_ADDRE
...
+        CodeVersionManager * pCodeVersionManager = pModule->GetCodeVersionManager();
+        ILCodeVersion activeILVersion = pCodeVersionManager->GetActiveILCodeVersion(pModule, methodTk);
+        *pOptimizationsDisabled = activeILVersion.IsDeoptimized();
+    }
```
Comment (hoyosjs) [inline src/coreclr/debug/daccess/dacdbiimpl.cpp:7424]:
```suggestion
    CodeVersionManager * pCodeVersionManager = pModule->GetCodeVersionManager();
    ILCodeVersion activeILVersion = pCodeVersionManager->GetActiveILCodeVersion(pModule, methodTk);
    *pOptimizationsDisabled = activeILVersion.IsDeoptimized();
```

You don't need `codeVersioningLockHolder` or the scope here.

Comment (davmason) [reply src/coreclr/debug/daccess/dacdbiimpl.cpp:7424]: You need to hold the lock to call `GetActiveILCodeVersion`

File: src/coreclr/debug/di/rsfunction.cpp [hunk 4]
```diff
@@ -673,27 +673,12 @@ HRESULT CordbFunction::AreOptimizationsDisabled(BOOL *pOptimizationsDisabled)
...
-    lockHolder.Acquire();
-
-    _ASSERTE(event.type == DB_IPCE_IS_OPTS_DISABLED_RESULT);
-    
-    *pOptimizationsDisabled = event.IsOptsDisabledData.value;
-
-    return event.hr;;
+    EX_TRY
+    {
+        hr = GetProcess()->GetDAC()->AreOptimizationsDisabled(GetModule(), GetMetadataToken(), pOptimizationsDisabled);
```
Comment (hoyosjs) [inline src/coreclr/debug/di/rsfunction.cpp:678]:
```suggestion
        hr = GetProcess()->GetDAC()->AreOptimizationsDisabled(GetModule()->GetRuntimeModule(), GetMetadataToken(), pOptimizationsDisabled);
```

File: src/coreclr/debug/daccess/dacdbiimpl.cpp (hunk 3 above)
Comment (hoyosjs) [reply src/coreclr/debug/daccess/dacdbiimpl.cpp:7424]: Not in the DAC

File: src/coreclr/debug/daccess/dacdbiimpl.cpp (hunk 1 above)
Comment (mikelle-rogers) [reply src/coreclr/debug/daccess/dacdbiimpl.cpp:7420]: I tested the latest commit, it works!

File: src/coreclr/debug/di/rsfunction.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (mikelle-rogers) [reply src/coreclr/debug/di/rsfunction.cpp]: @hoyosjs

File: src/coreclr/debug/daccess/dacdbiimpl.cpp (hunk 3 above)
Comment (hoyosjs) [reply src/coreclr/debug/daccess/dacdbiimpl.cpp]: I don't know what mechanism is in place though to ensure the DAC doesn't pick a code version while a BG jit happens in the target process.

File: src/coreclr/debug/di/rsfunction.cpp [hunk 5]
```diff
@@ -673,27 +673,12 @@ HRESULT CordbFunction::AreOptimizationsDisabled(BOOL *pOptimizationsDisabled)
...
-    lockHolder.Acquire();
-
-    _ASSERTE(event.type == DB_IPCE_IS_OPTS_DISABLED_RESULT);
-    
-    *pOptimizationsDisabled = event.IsOptsDisabledData.value;
-
-    return event.hr;;
+    EX_TRY
+    {
+        hr = GetProcess()->GetDAC()->AreOptimizationsDisabled(GetModule()->GetRuntimeModule(), m_MDToken, pOptimizationsDisabled);
```
Comment (hoyosjs) [inline src/coreclr/debug/di/rsfunction.cpp:678]:
```suggestion
        hr = GetProcess()->GetDAC()->AreOptimizationsDisabled(GetModule()->GetRuntimeModule(), GetMetadataToken(), pOptimizationsDisabled);
```

File: src/coreclr/debug/ee/debugger.cpp [hunk 6]
```diff
@@ -10468,34 +10468,6 @@ bool Debugger::HandleIPCEvent(DebuggerIPCEvent * pEvent)
         }
         break;
 
-    case DB_IPCE_IS_OPTS_DISABLED:
```
Comment (hoyosjs) [inline src/coreclr/debug/ee/debugger.cpp:10471]: You can also remove all the definitions of `DB_IPCE_IS_OPTS_DISABLED` and `DB_IPCE_IS_OPTS_DISABLED_RESULT` and also `IsOptsDisabledData`

File: src/coreclr/debug/shared/dbgtransportsession.cpp [hunk 7]
```diff
@@ -2499,13 +2499,6 @@ DWORD DbgTransportSession::GetEventSize(DebuggerIPCEvent *pEvent)
         break;
 
     case DB_IPCE_DISABLE_OPTS:
-    case DB_IPCE_IS_OPTS_DISABLED:
```
Comment (hoyosjs) [inline src/coreclr/debug/shared/dbgtransportsession.cpp:2502]: You can also remove `DB_IPCE_IS_OPTS_DISABLED` from `src/coreclr/debug/inc/dbgipceventtypes.h`

**Review by hoyosjs — APPROVED** [review]
Other than that LGTM

**Pull request merged**